### PR TITLE
workspace.onDidRenameFile and workspace.onWillRenameFile implementati…

### PR DESCRIPTION
**Pull request by vrubezhny**
…ons are missed #4167

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

File: packages/plugin-ext/src/plugin/workspace.ts [hunk 1]
```diff
@@ -276,4 +276,17 @@ export class WorkspaceExtImpl implements WorkspaceExt {
         return normalize(result, true);
     }
 
+    // Experimental API https://github.com/theia-ide/theia/issues/4167
+    private workspaceWillRenameFileEmitter = new Emitter<theia.FileWillRenameEvent>();
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/workspace.ts]: Who is firing them? Otherwise they just stubs

**Comment by benoitf** [comment]
hello, same remark than @akosyakov Did you forget some other files in your commit ? Emitters are not bound

**Comment by vrubezhny** [comment]
@akosyakov @benoitf I've added a fixup that adds "file rename" and "will rename" events support, so now these events should be fired on file rename operations

**Comment by akosyakov** [comment]
@vrubezhny How do you test it?

**Comment by vrubezhny** [comment]
> @vrubezhny How do you test it?

In my local environment there were added additional lines that print out the messages on events firing as well as on listeners set up/invocation.
So, when MS Typescript extension plugin installed (https://github.com/che-incubator/ms-code.typescript/releases/download/v1.30.2/che-typescript-language.vsix), on startup I was informed when it sets its listener for `FileRenameEvent` (it looks like it doesn't use `FileWillRenameEvent`) and then I was playing with renaming of *.ts files in ma test project and watching for the according messages from the listeners.

**Comment by akosyakov** [comment]
@vrubezhny with this changes and https://github.com/che-incubator/ms-code.typescript/releases/download/v1.30.2/che-typescript-language.vsix renaming a file should update imports right?  I would try

**Comment by vrubezhny** [comment]
Yes it should and it updates imports... but sometimes there are issues (a path to import might be updated incorrectly). I believe it's an issue in  MS Typescript, not in workspace.onDidRenameFile and workspace.onWillRenameFile implementations.

ondidrenamefile

**Comment by vrubezhny** [comment]
> It works but very slow about 10-15s till TS extension asks to rename imports. We need to investigate separately where time is lost.
> 
> **Update**
> 
> Although PR is not rebased, maybe on master state, it is better. -> It's the same, please open an issue.

I believe it is a limitation of FileSystemWatcher: 

https://github.com/theia-ide/theia/blob/master/packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.spec.ts#L36

**Comment by vrubezhny** [comment]
I have squashed and rebased the commits.

**Comment by akosyakov** [comment]
> I believe it is a limitation of File System Watcher:

We fire move events eagerly. There should not be any delays.

**Pull request merged**